articles/ros2_dds_security.md

```diff
@@ -0,0 +1,220 @@
+
+RCL supports finding security files in one directory that is the root of a directory structure corresponding to the fully-qualified names of every node instance (i.e. namespace + node name).
+For example, for the `/front/camera` node, the directory structure would look like:
+
+    <root>
+    └── front
+        └── camera
+            ├── cert.pem
+            ├── key.pem
+            ├── ...
+
+To be clear: this directory structure needs to reflect the state of the running system.
+In other words, it does not contain a set of files per node on disk, but per node instance _in the ROS graph_.
+
+This can be specified by setting the `$ROS_SECURITY_ROOT_DIRECTORY` environment variable to point to the root of the directory tree.
+The set of files expected within each node instance directory are the same as outlined in the "Manual specification" section, above.
+
+
+##### Support security files lookup methods
+
+If using the directory tree approach to organize security files, RCL supports two different methods for looking up a given node instance's security files in the tree:
+
+- **Exact**: Only load security files from a directory exactly matching the fully-qualified name of the node instance.
+For example, given a node named "baz" within the "/foo/bar/" namespace, only load security files from `<root>/foo/bar/baz/`.
+This is the default behavior.
+- **Prefix**: Attempt to load the most specific set of security files, but if they can't be found, check for security files under a less-specific node name.
+For example, given a node named "baz_123" within the "/foo/bar/" namespace, load security files from `<root>/foo/bar/baz_123/`.
+However, if that directory doesn't exist, find the most specific (i.e. longest) node name that _does_ have security files within that namespace (e.g. `<root>/foo/bar/baz_12/`, or `<root>/foo/bar/baz/`, etc.).
+Note that it will not search higher in the namespace hierarchy.
+
+The desired lookup method can be specified by setting the `$ROS_SECURITY_LOOKUP_TYPE` environment variable to "MATCH_EXACT" (case-sensitive) for the **Exact* method, or "MATCH_PREFIX" (case-sensitive) for the **Prefix** method.
+
+
+### Support for both permissive and strict enforcement of security
+
+Nodes with the security features enabled will not communicate with nodes that don't, but what should RCL do if one tries to launch a node that has no discernable keys/permissions/etc.? It has two options:
+
+- **Permissive mode**: Try to find security files, and if they can't be found, launch the node without enabling any security features.
+This is the default behavior.
+- **Strict mode**: Try to find security files, and if they can't be found, fail to run the node.
+
+The type of mode desired can be specified by setting the `$ROS_SECURITY_STRATEGY` environment variable to "Enforce" (case-sensitive) for strict mode, and anything else for permissive mode.
+
+
+### Support for a master "on/off" switch for all SROS 2 features
```

Comment:
I considered that, and I'm still kinda on the fence. In one sense it makes sense to show the simplest thing first, but the reason I put it here is so folks have context and understand what exactly is being disabled by the time they get to it (and then I can say "all the above stuff will be ignored" essentially, instead of being more abstract). I don't feel strongly about it, but that's my reasoning. If you'd like it moved, let me know.

Reply:
I like it as-is.

Reply:
> the reason I put it here is so folks have context and understand what exactly is being disabled by the time they get to it (and then I can say "all the above stuff will be ignored" essentially, instead of being more abstract)

Fine by me :+1: